### PR TITLE
Some stub code for the UA

### PR DESCRIPTION
**Pull request by martinthomson**
I've started to build out some of the code here for the UA side of things (to start at the beginning).

I haven't quite worked out how this is going to work for helpers yet, but I've some idea of the layout of the user piece.

The idea is that users are identified with numbers (0 through n).  The tool accepts a range of numbers when identifying users and operates on all of those users at once.  I have two functions so far: setup (which needs helper public keys, so it doesn't work just yet) and set-match-key (which doesn't so it should be OK).

I'm assuming that all state will be serialized to individual JSON files, one for each user.

I'm guessing that the helpers will have two files each: a public file with all their public keys, and a private file with all of the same information, plus the secrets and other state they need to operate.  This includes a wrapper that pulls in all the public helper information into the one struct.

I think that the helpers will each be different, so I've created 4 stub classes named Helper1 through Helper4.  It would be really good to get better names for these. I'd also like names for the first pair and the second pair.

Depends on #4.

### CODE REVIEW

**Comment by martinthomson** [comment]
To start out, you need to run setup for helpers and the UAs.

This will setup all the helpers:

```sh
cargo run --bin raw-ipa-helper -- -vvvv setup seh
cargo run --bin raw-ipa-helper -- -vvvv setup teh
cargo run --bin raw-ipa-helper -- -vvvv setup ah1
cargo run --bin raw-ipa-helper -- -vvvv setup ah2
```

This will setup users 0 through 9:
```sh
cargo run --bin raw-ipa-ua -- -vvvv setup -u 0..10
```

This will set matchkeys for user 3:
```sh
cargo run --bin raw-ipa-ua -- -vvvv  set-match-key example.com 1234123412341234123412341234123412341234123412341234123412341234 -u 3
```

Obviously you want different matchkeys for each user, but you can put in a range of users if you want.

The argument order is a little annoying, but I've done what I can there.

All of these create JSON files in `./db`, which you can look at to see what is going on.

File: raw-ipa-lib/src/user.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    #[test]
+    fn matchkey_random() {
+        let mut rng = thread_rng();
+        let d1 = ThresholdDecryptionKey::new(&mut rng);
+        let d2 = ThresholdDecryptionKey::new(&mut rng);
+        let tek = ThresholdEncryptionKey::new(&[d1.encryption_key(), d2.encryption_key()]);
+        let u = User::new(0, tek);
+
```
Comment (benjaminsavage) [inline raw-ipa-lib/src/user.rs:132]: maybe add a comment here about how we are NOT setting a match key. Took me a while to spot the difference with the test above.

File: raw-ipa-lib/src/user.rs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        let complete1 = d2.decrypt(partial1);
+
+        assert_ne!(complete1, User::point_from_matchkey(MATCHKEY));
+
+        // A second matchkey is completely random and different again.
+        let c = u.encrypt_matchkey(ORIGIN);
+        let partial2 = d1.threshold_decrypt(c);
+        let complete2 = d2.decrypt(partial2);
+        assert_ne!(complete1, complete2);
+        assert_ne!(complete2, User::point_from_matchkey(MATCHKEY));
```
Comment (benjaminsavage) [inline raw-ipa-lib/src/user.rs:144]: So actually, I think this test *should fail*. This relates to my other comment about defaulting to same-device attribution working when no match key is set for that domain.

**Review by benjaminsavage — COMMENTED** [review]
Just some general questions and feedback.

File: raw-ipa-lib/src/user.rs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+#[cfg(test)]
+mod tests {
+    use super::User;
+    use crate::threshold::{
+        DecryptionKey as ThresholdDecryptionKey, EncryptionKey as ThresholdEncryptionKey,
+    };
+    use rand::thread_rng;
+
+    const MATCHKEY: &[u8; 32] = &[0; 32];
```
[the remark this answers is not in the document]
Comment (martinthomson) [reply raw-ipa-lib/src/user.rs:125]: It is currently valid :)  I don't see much point in making this unique, though it has some unfortunate properties...

File: raw-ipa-lib/src/user.rs (hunk 2 above)
Comment (martinthomson) [reply raw-ipa-lib/src/user.rs]: ACK, yes.

File: raw-ipa-lib/src/user.rs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    /// # Errors
+    /// When the file cannot be written.
+    #[cfg(feature = "enable-serde")]
+    pub fn save(&self, dir: &Path) -> Res<()> {
+        let f = self.filename(dir);
+        fs::write(f, serde_json::to_string(self)?.as_bytes())?;
+        Ok(())
+    }
+
+    fn point_from_matchkey(mk: &[u8; 32]) -> RistrettoPoint {
```
[the remark this answers is not in the document]
Comment (benjaminsavage) [reply raw-ipa-lib/src/user.rs:71]: What do you imagine the eventual JavaScript API might require of match keys? Are you envisioning similarly requiring the caller to provide 32 bytes of binary?

File: raw-ipa-lib/src/user.rs [hunk 5]
```diff
@@ -64,30 +70,47 @@ impl User {
...
+                );
+                info.extend_from_slice(p_bytes);
+                let mut mk = [0; 32];
+                Hkdf::<Sha512>::from_prk(&self.fallback_prk)
+                    .unwrap() // prk came from Hkdf<Sha512>
+                    .expand(&info, &mut mk)
+                    .unwrap(); // length is valid
+                let m = Self::point_from_matchkey(&mk);
+                self.threshold_key.encrypt(m, &mut thread_rng())
+            });
```
Comment (benjaminsavage) [inline raw-ipa-lib/src/user.rs:112]: Wow. That's a lot more complex than what I had imagined would be necessary - but I'm sure there's a good reason for all of this I don't understand.

I had imagined we would just call something like:
```
self.encrypted_match_keys
    .insert(String::from(provider), RistrettoPoint::random(&mut rng));
```

File: raw-ipa-lib/src/user.rs [hunk 6]
```diff
@@ -64,30 +70,47 @@ impl User {
 
     fn point_from_matchkey(mk: &[u8; 32]) -> RistrettoPoint {
         // Note that ristretto wants 64 bytes of input; also we don't know if the input is uniform.
-        // TODO: Consider salting this input somehow (with the origin, perhaps).
+        // TODO: Consider salting this input somehow (with the provider, perhaps).
```
[the remark this answers is not in the document]
Comment (benjaminsavage) [reply raw-ipa-lib/src/user.rs:73]: Is this related to my final comment below - about default match key from provider1 matching with default key from provider2?

File: raw-ipa-lib/src/user.rs [hunk 7]
```diff
@@ -122,25 +146,52 @@ mod tests {
...
+    fn matchkey_fallback() {
+        let mut rng = thread_rng();
+        let d1 = ThresholdDecryptionKey::new(&mut rng);
+        let d2 = ThresholdDecryptionKey::new(&mut rng);
+        let tek = ThresholdEncryptionKey::new(&[d1.encryption_key(), d2.encryption_key()]);
+        let u = User::new(0, tek);
+        // no matchkeys set here
+
+        let c1 = u.encrypt_matchkey(PROVIDER);
+        let c2 = u.encrypt_matchkey(PROVIDER);
```
[the remark this answers is not in the document]
Comment (benjaminsavage) [reply raw-ipa-lib/src/user.rs:185]: You're right - I missed that. Thanks for adding the test!

File: raw-ipa-lib/src/user.rs [hunk 8]
```diff
@@ -33,10 +35,14 @@ impl User {
     /// When a file for the given ID already exists.
     #[must_use]
     pub fn new(id: usize, threshold_key: ThresholdEncryptionKey) -> Self {
+        let mut ikm = [0; 64];
```
[the remark this answers is not in the document]
Comment (benjaminsavage) [reply raw-ipa-lib/src/user.rs:38]: That is kind of magical... I'm not used to languages that look beyond the current line, to see how it's subsequently used to infer the type, that's really amazing.

**Pull request merged**